### ISSUE TITLE
Inconsistency when using DataLoader

### ISSUE THREAD

**Issue by davidsonic** [rusty1s/pytorch_geometric#437]
## ❓ Questions & Help

I encountered a weird problem during using the library's provided dataloader and haven't figured out whether it's because of InMemoryDataset or DataLoader provided. I created a dataset inherited from InMemoryDataset. 
The first two graph looks like this:
![image](https://user-images.githubusercontent.com/15846113/59902977-8e29a100-93b4-11e9-9a6f-018877c63d16.png)
If I set shuffle=False and iterate twice with batch_size=1, I get what I want:
![image](https://user-images.githubusercontent.com/15846113/59903022-be713f80-93b4-11e9-878f-2bfc770516fb.png)
However, if I set batch_size=2 and iterate once, the result is not consistent and some doesn't even appear in the dataset:
![image](https://user-images.githubusercontent.com/15846113/59903094-f082a180-93b4-11e9-8825-17c8df770554.png)
![image](https://user-images.githubusercontent.com/15846113/59903109-fb3d3680-93b4-11e9-8aa9-c664ed0b06eb.png)
From the above two pics, it's obvious that ids (created for debugging)  don't correspond to edge_index. Why?

**Comment by rusty1s** [comment]
Which attributes are corrupted? `edge_item_index`? This might be very well because we cumulatively sum up attributes which contain the words `index` or `face` in mini-batch scenarios. You can rename your data attribute to, *e.g.*, `edge_item`, or inherit from the `Data` class and redefine [`__cumsum__`](https://rusty1s.github.io/pytorch_geometric/build/html/_modules/torch_geometric/data/data.html#Data.__cumsum__). Sorry for any inconvenience.

**Comment by davidsonic** [comment]
Thanks for your reply! I changed the attribute as suggested to edge_item and got an error during collating. The attribute is of same dimension as edge_index (2x#edges), serves as id for edges for debugging purpose. Isn't it supposed to be supported as an additional attribute for Data class. 
![image](https://user-images.githubusercontent.com/15846113/59905844-0f386680-93bc-11e9-9541-a4a564b68dc7.png)
![image](https://user-images.githubusercontent.com/15846113/59905800-f9c33c80-93bb-11e9-80ed-4a92a54447df.png)

**Comment by rusty1s** [comment]
Ah I see! Man, that sucks. It's because PyG doesn't know in which dimension it should concatenate the individual tensors (we concatenate everything in the first dimension, expect it is an `index` tensor). Ok, so the best approach is simply to create a new data class and customize mini-batch behavior:

```python
from torch_geometric.data import Data

class CustomData(Data):
    def __init__(**kwargs):
        super(CustomData, self).__init__(**kwargs)

    def __cat_dim__(self, key, value):
        return -1 if key in ['edge_index', 'edge_item'] else 0

    def __cumsum__(self, key, value):
        return key == 'edge_index'
```

**Comment by davidsonic** [comment]
Thanks for the confirmation! Close it for now.
